storagemarketadapter/provider_adapter.go

```diff
@@ -0,0 +1,196 @@
+package storagemarketadapter
+
+// this file implements storagemarket.StorageProviderNode
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+	unixfile "github.com/ipfs/go-unixfs/file"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/lib/padreader"
+	"github.com/filecoin-project/lotus/node/modules/dtypes"
+	"github.com/filecoin-project/lotus/storage/sectorblocks"
+	"github.com/filecoin-project/lotus/storagemarket"
+)
+
+var log = logging.Logger("provideradapter")
+
+type ProviderNodeAdapter struct {
+	api.FullNode
+
+	// this goes away with the data transfer module
+	dag dtypes.StagingDAG
+
+	secb *sectorblocks.SectorBlocks
+}
+
+func NewProviderNodeAdapter(dag dtypes.StagingDAG, secb *sectorblocks.SectorBlocks, full api.FullNode) storagemarket.StorageProviderNode {
+	return &ProviderNodeAdapter{
+		FullNode: full,
+		dag:      dag,
+		secb:     secb,
+	}
+}
+
+func (n *ProviderNodeAdapter) PublishDeals(ctx context.Context, deal storagemarket.MinerDeal) (storagemarket.DealID, cid.Cid, error) {
+	log.Info("publishing deal")
+
+	worker, err := n.StateMinerWorker(ctx, deal.Proposal.Provider, nil)
+	if err != nil {
+		return 0, cid.Undef, err
+	}
+
+	params, err := actors.SerializeParams(&actors.PublishStorageDealsParams{
+		Deals: []actors.StorageDealProposal{deal.Proposal},
+	})
+
+	if err != nil {
+		return 0, cid.Undef, xerrors.Errorf("serializing PublishStorageDeals params failed: ", err)
+	}
+
+	// TODO: We may want this to happen after fetching data
+	smsg, err := n.MpoolPushMessage(ctx, &types.Message{
+		To:       actors.StorageMarketAddress,
+		From:     worker,
+		Value:    types.NewInt(0),
+		GasPrice: types.NewInt(0),
+		GasLimit: types.NewInt(1000000),
+		Method:   actors.SMAMethods.PublishStorageDeals,
+		Params:   params,
+	})
+	if err != nil {
+		return 0, cid.Undef, err
+	}
+	r, err := n.StateWaitMsg(ctx, smsg.Cid())
```

Comment:
We'll also need to wait for some finality here, and handle chain reverts (otherwise the dealID can change)